src/backend/distributed/planner/multi_master_planner.c

```diff
@@ -203,10 +257,43 @@ BuildSelectStatement(Query *masterQuery, List *masterTargetList, CustomScan *rem
 		topLevelPlan = &remoteScan->scan.plan;
 	}
 
+	/*
+	 * distinct on() requires sort + unique plans. If already has order by
+	 * clause we append distinct clauses to the end of it. Postgresql requires
+	 * that if both distinct on() and order by exists, ordering is first done
+	 * on distinct clauses. Therefore we can safely append distinct clauses to
+	 * the end of order by clauses. Although the same column may appear more
+	 * than once in order by clauses, created plan uses only one instance, for
+	 * example order by a,b,a,a,b,c is translated to equivalent order by a,b,c.
+	 *
+	 * If the query has distinct clause but not distinct on, we first create
+	 * distinct plan that is either HashAggreate or Sort + Unique plans depending
+	 * on hashable property of columns in distinct clause. If there is order by
+	 * clause, it is handled after distinct planning.
+	 */
+	if (masterQuery->hasDistinctOn)
+	{
+		ListCell *distinctCell = NULL;
+		foreach(distinctCell, masterQuery->distinctClause)
+		{
+			SortGroupClause *singleDistinctClause = lfirst(distinctCell);
+			Index sortGroupRef = singleDistinctClause->tleSortGroupRef;
+
+			if (!get_sortgroupref_clause_noerr(sortGroupRef, sortClauseList))
```

Comment:
`get_sortgroupref_clause_noerr()` returns a pointer. Shall we do a NULL check instead of `!`?

Reply:
ok, replaced with null check